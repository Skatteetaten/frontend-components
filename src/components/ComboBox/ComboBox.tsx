import * as React from 'react';
import {
  VirtualizedComboBox as FabricComboBox,
  IComboBoxProps
} from 'office-ui-fabric-react/lib-commonjs/ComboBox';
import Callout from '../Callout/Callout';
import classnames from 'classnames';

import { Label } from 'office-ui-fabric-react/lib-commonjs/Label';
import { IconButton } from 'office-ui-fabric-react/lib-commonjs/Button';
import {
  getClassNames,
  getErrorClassNames,
  getLabelClassNames,
  getOptionsClassNames
} from './ComboBox.classNames';

// This is just a workaround so that the combobox opens on focus.
class SkeCombobox extends FabricComboBox {
  focus = () => {
    if (
      // @ts-ignore TODO
      this.props.expandOnFocus === true &&
      // @ts-ignore TODO
      this._comboBox.value &&
      this.props.disabled === false
    ) {
      // @ts-ignore TODO
      this._comboBox.value.focus(this.props.shouldOpenOnFocus);
      return true;
    }
    return false;
  };
}

interface ComboboxProps extends IComboBoxProps {
  /** Angir om valgene automatisk skal vises når feltet har focus */
  expandOnFocus?: boolean;
  /** Hjelpetekst */
  help?: JSX.Element | string;
  /** Størrelse på inputfelt som skal benyttes */
  inputSize?: 'normal' | 'large';
}
type ComboboxState = {
  isCalloutVisible: boolean;
};
/**
 * @visibleName ComboBox (Nedtrekksliste med skriving)
 */
export default class Combobox extends React.PureComponent<
  ComboboxProps,
  ComboboxState
> {
  static defaultProps = {
    autoComplete: 'on',
    expandOnFocus: false,
    allowFreeform: false,
    label: undefined,
    errorMessage: undefined,
    help: undefined,
    disabled: false,
    shouldOpenOnFocus: true
  };
  // @ts-ignore TODO
  constructor(props) {
    super(props);
    this.state = {
      isCalloutVisible: false
    };
    this._onRenderLabel = this._onRenderLabel.bind(this);
    this._onClick = this._onClick.bind(this);
    this._onDismiss = this._onDismiss.bind(this);
  }
  // @ts-ignore TODO
  _onRenderLabel(props) {
    const { label, help, componentId } = props;

    const styles = getLabelClassNames(props);
    let { isCalloutVisible } = this.state;
    const helpElement = React.isValidElement(help) ? help : <p>{help}</p>;
    return (
      <div className={styles.labelArea}>
        <span className={styles.label}>
          {label ? (
            <Label className={styles.labelText} htmlFor={componentId}>
              {label}
            </Label>
          ) : null}
        </span>
        {help && (
          <span
            className={styles.labelIconArea}
            // @ts-ignore TODO
            ref={menuButton => (this._iconButtonElement = menuButton)}
          >
            <IconButton
              iconProps={{ iconName: 'HelpOutline' }}
              title="HelpOutline"
              ariaLabel="Hjelp"
              onClick={this._onClick}
              className={styles.icon}
              aria-haspopup={true}
            />
          </span>
        )}
        {isCalloutVisible && (
          <Callout
            directionalHint={Callout.POS_TOP_LEFT}
            color={Callout.HELP}
            ariaLabel={'Hjelpetekst'}
            // @ts-ignore TODO
            target={this._iconButtonElement}
            onClose={this._onDismiss}
          >
            {helpElement}
          </Callout>
        )}
      </div>
    );
  }

  _onClick() {
    this.setState({
      isCalloutVisible: !this.state.isCalloutVisible
    });
  }

  _onDismiss() {
    this.setState({
      isCalloutVisible: false
    });
  }

  render() {
    const {
      children,
      errorMessage,
      label,
      help,
      // @ts-ignore TODO
      onRenderLabel,
      className,
      id,
      ...props
    } = this.props;
    const labelProps = { label, help };
    return (
      <div id={id}>
        {onRenderLabel
          ? onRenderLabel(labelProps)
          : this._onRenderLabel(labelProps)}
        <SkeCombobox
          {...props}
          role="combobox"
          className={classnames(getClassNames(this.props), className)}
          onFocus={this._basicComboBoxOnClick}
          componentRef={this._basicComboBoxComponentRef}
          calloutProps={{
            className: getOptionsClassNames(this.props)
          }}
        >
          {children}
        </SkeCombobox>
        {errorMessage && (
          <label className={getErrorClassNames(this.props)}>
            {errorMessage}
          </label>
        )}
      </div>
    );
  }
  // @ts-ignore TODO
  _basicComboBoxComponentRef = component => {
    // @ts-ignore TODO
    this._basicCombobox = component;
  };

  _basicComboBoxOnClick = () => {
    // @ts-ignore TODO
    this._basicCombobox.focus();
    const { onFocus } = this.props;
    // @ts-ignore TODO
    onFocus && onFocus();
  };
}
