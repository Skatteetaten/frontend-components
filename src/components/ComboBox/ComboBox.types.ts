import { IComboBoxProps, ICalloutProps } from '@fluentui/react';
import { calloutState, LabelWithCalloutProps } from '../LabelWithCallout';

export interface ComboBoxProps extends IComboBoxProps {
  /** Egendefinert feilmelding */
  errorMessage?: IComboBoxProps['errorMessage'];
  /** Hjelpetekst */
  help?: JSX.Element | string;
  /** Størrelse på inputfelt som skal benyttes */
  inputSize?: 'normal' | 'large';
  /** aria-label for knapp i label */
  labelButtonAriaLabel?: string;
  /** Om feltet er obligatorisk */
  required?: boolean;
  /** Overstyr label, se LabelWithCallout komponent */
  labelWithCalloutProps?: LabelWithCalloutProps;
  /** Brukerspesifisert event for callout */
  onCalloutToggle?: (
    oldCalloutState: calloutState,
    newCalloutState: calloutState
  ) => void;
  /** CalloutProps som sendes videre ned til Callout */
  calloutProps?: ICalloutProps;
  /** Lesemodus. Kan brukes i sammenheng med text eller defaultSelectedKey for å vise verdi */
  readOnly?: boolean;
  /** Om feltet er obligatorisk og skal markeres med stjerne (*) */
  requiredWithMark?: boolean;
}
