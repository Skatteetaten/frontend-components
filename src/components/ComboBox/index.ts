export { default } from './ComboBox';
export * from './ComboBox';
