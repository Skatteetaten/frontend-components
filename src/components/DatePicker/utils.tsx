export const DEFAULTFORMATDATE = (date: Date | null | undefined): string => {
  return date
    ? Intl.DateTimeFormat('no', {
        year: 'numeric',
        month: '2-digit',
        day: 'numeric',
      }).format(date)
    : '';
};

export const DEFAULTPARSEDATEFROMSTRING = (date: string): Date | null => {
  try {
    if (date && date.length === 10) {
      const day: number = parseInt(date.substring(0, 2));
      const month: number = parseInt(date.substring(3, 5));
      const year: number = parseInt(date.substring(6, 10));
      const formattedDate: Date = new Date(Date.UTC(year, month - 1, day));
      return formattedDate;
    }
  } catch (error) {
    console.log(error);
  }
  return null;
};

export const monthsForLocale = (
  localeName = 'no',
  monthFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { month: monthFormat })
    .format;
  const months: string[] = [];
  for (let month = 0; month < 12; month++) {
    months.push(format(new Date(Date.UTC(2020, month))));
  }
  return months;
};

export const weekdaysForLocale = (
  localeName = 'no',
  dayFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { weekday: dayFormat })
    .format;
  const weekdays: string[] = [];
  for (let day = 0; day < 7; day++) {
    weekdays.push(format(new Date(Date.UTC(2020, 7, day + 2)))); // add 2 days to let the week start on Sunday
  }
  return weekdays;
};
