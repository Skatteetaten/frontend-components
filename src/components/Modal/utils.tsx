import { useEffect } from 'react';

export const createModalDomPlacement = (
  setFlag: (a: boolean) => void,
  shadowRootNode: Document | ShadowRoot | undefined
) => {
  if (!shadowRootNode) {
    setFlag(true);
    return;
  }

  const styleInjectorWrapperId = 'style-injector';
  const modalWrapperId = 'modal-wrapper';

  if (shadowRootNode.getElementById(modalWrapperId)) {
    setFlag(true);
  } else {
    const modalWrapper = document.createElement('div');
    modalWrapper.id = modalWrapperId;
    const styleInjectorWrapper = shadowRootNode.getElementById(
      styleInjectorWrapperId
    );
    const parent = styleInjectorWrapper ? styleInjectorWrapper : shadowRootNode;
    parent.appendChild(modalWrapper);
    setFlag(true);
  }
};

export const useEscOnPress = (
  ref: HTMLDivElement | null,
  onClose?: (ref: HTMLDivElement) => void
): void => {
  useEffect(() => {
    const onKeyDown = (event: KeyboardEvent): void => {
      if (ref == null) {
        return;
      }
      if (event.code === 'Escape' && onClose) {
        onClose(ref);
      }
    };

    if (ref) {
      window.addEventListener('keydown', onKeyDown as EventListener, false);
    }
    return function cleanup(): void {
      if (window) {
        window.removeEventListener('keydown', onKeyDown as EventListener);
      }
    };
  }, [ref, onClose]);
};
