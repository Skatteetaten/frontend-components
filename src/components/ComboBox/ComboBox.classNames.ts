import { mergeStyles, mergeStyleSets } from '@uifabric/merge-styles';
import { getTheme } from '@uifabric/styling';
import { FontSizes, FontWeights } from '../utils/fonts';
import { MdIcons } from '../utils/icons/';
import { Animation } from '../utils/getAnimationStyles';
import { PaletteProps } from '..';
// @ts-ignore TODO
function getFieldTypeStyles(props) {
  const palette = getTheme().palette as PaletteProps;

  if (props.inputSize === 'large') {
    return {
      '& .ms-ComboBox': {
        borderWidth: '2px',
        borderRadius: '0px',
        height: '46px',
        padding: '5px 12px',
        borderColor: palette.skeColor.blackAlt
      },
      '& .ms-ComboBox-Input': {
        fontSize: FontSizes.large
      },
      '& button.ms-ComboBox-CaretDown-button': {
        top: '5px'
      },
      'i.ms-Button-icon': {
        fontSize: FontSizes.large
      }
    };
  } else {
    return {
      '& .ms-ComboBox-Input': {
        fontSize: FontSizes.small
      }
    };
  }
}
// @ts-ignore TODO
export const getClassNames = props => {
  const { errorMessage } = props;
  const palette = getTheme().palette as PaletteProps;
  const color = errorMessage
    ? palette.skeColor.error
    : palette.skeColor.blackAlt;

  return mergeStyles({
    displayName: 'SkeCombobox',
    selectors: {
      '.ms-ComboBox': errorMessage && {
        borderColor: color
      },
      '& .ms-ComboBox': {
        borderRadius: '0px',
        height: '30px',
        padding: '1px 32px 1px 12px',
        borderColor: palette.skeColor.blackAlt
      },
      '& .ms-ComboBox:active, & .ms-ComboBox:focus, & .ms-ComboBox::after, &.is-active.ms-ComboBox, .ms-ComboBox.is-open': {
        border: `2px solid ${palette.skeColor.blue}`
      },
      '& .ms-ComboBox-Input': {
        paddingBottom: '1px',
        paddingTop: '1px'
      },
      '& .ms-ComboBox-Input.is-disabled': {
        backgroundColor: palette.skeColor.whiteGrey
      },
      '& .ms-ComboBox.is-disabled': {
        borderColor: palette.skeColor.grey
      },
      '& .ms-ComboBox.is-disabled button': {
        color: palette.skeColor.grey
      },
      '.ms-ComboBox-CaretDown-button': {
        // Negative positioning to account for the 2px border
        right: '-2px',
        top: '-2px'
      },
      ...getFieldTypeStyles(props)
    }
  });
};
// @ts-ignore TODO
export const getErrorClassNames = props => {
  const { errorMessage } = props;
  const palette = getTheme().palette as PaletteProps;
  const color = errorMessage
    ? palette.skeColor.error
    : palette.skeColor.blackAlt;
  const errorIcon = "'" + MdIcons.icons.Error + "'";

  return mergeStyles(Animation.errorMessage, {
    displayName: 'SkeComboBoxError',
    color,
    fontSize: FontSizes.small,
    fontWeight: '400',
    paddingTop: '5px',
    display: 'flex',
    alignItems: 'center',
    paddingLeft: 20,
    position: 'relative',
    selectors: {
      '&:before': {
        fontFamily: MdIcons.fontFace.fontFamily,
        fontSize: 18,
        display: 'block',
        content: errorIcon,
        marginRight: 3,
        position: 'absolute',
        top: 5,
        left: 0
      }
    }
  });
};
// @ts-ignore TODO
export const getOptionsClassNames = props => {
  const palette = getTheme().palette as PaletteProps;

  return mergeStyles({
    displayName: 'SkeComboBoxOptions',
    selectors: {
      '& .ms-ComboBox-option.is-checked': {
        border: '2px solid' + palette.skeColor.blue,
        outline: 'none'
      },
      '& .ms-ComboBox-option.is-checked:after': {
        border: 'none',
        outline: 'none'
      }
    }
  });
};
// @ts-ignore TODO
export const getLabelClassNames = props => {
  const palette = getTheme().palette as PaletteProps;

  return mergeStyleSets({
    labelArea: {
      display: 'flex',
      alignItems: 'center'
    },
    label: {
      //flexGrow: 1,
    },
    labelText: {
      fontSize: FontSizes.small,
      color: palette.skeColor.blackAlt,
      fontWeight: FontWeights.regular
    },
    labelIconArea: {
      height: '27px',
      marginTop: '-5px'
    },
    icon: {
      color: palette.skeColor.blue,
      selectors: {
        '& i': {
          fontSize: 'large'
        },
        '&:focus&:after': {
          border: `2px solid ${palette.skeColor.blue}`,
          outline: 'none'
        },
        '&:focus::after': {
          border: `2px solid ${palette.skeColor.blue}`,
          outline: 'none'
        },
        '&:hover': {
          border: palette.skeColor.blackAlt,
          outline: 'none',
          background: 'none'
        }
      }
    }
  });
};
