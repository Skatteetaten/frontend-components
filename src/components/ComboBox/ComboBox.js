import React, { isValidElement } from 'react';
import { VirtualizedComboBox as FabricComboBox } from 'office-ui-fabric-react/lib-commonjs/ComboBox';
import Callout from '../Callout/Callout';
import classnames from 'classnames';

import { Label } from 'office-ui-fabric-react/lib-commonjs/Label';
import { IconButton } from 'office-ui-fabric-react/lib-commonjs/Button';
import {
  getClassNames,
  getErrorClassNames,
  getLabelClassNames,
  getOptionsClassNames
} from './ComboBox.classNames';
import PropTypes from 'prop-types';

// This is just a workaround so that the combobox opens on focus.
class SkeCombobox extends FabricComboBox {
  focus = () => {
    if (
      this.props.expandOnFocus === true &&
      this._comboBox.value &&
      this.props.disabled === false
    ) {
      this._comboBox.value.focus(this.props.shouldOpenOnFocus);
      return true;
    }
    return false;
  };
}

/**
 * @visibleName ComboBox (Nedtrekksliste med skriving)
 */
export default class Combobox extends React.PureComponent {
  static propTypes = {
    /** Angir om fullføring av teksten skal være aktiv eller ikke */
    autoComplete: PropTypes.oneOf(['on', 'off']),
    /** Angir om valgene automatisk skal vises når feltet har focus */
    expandOnFocus: PropTypes.bool,
    /** Angir om brukeren skal kunne skrive elementer som ikke er forhåndsdefinert */
    allowFreeform: PropTypes.bool,
    /** Beskrivelse over feltet */
    label: PropTypes.string,
    /** Hjelpetekst */
    help: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
    /** Feilmelding under feltet */
    errorMessage: PropTypes.string,
    /** Inaktiv eller ikke */
    disabled: PropTypes.bool,
    className: PropTypes.string,
    /** Global attributt som må være unik for hele HTML dokumentet */
    id: PropTypes.string,
    /** Størrelse på inputfelt som skal benyttes */
    inputSize: PropTypes.oneOf(['normal', 'large'])
  };

  static defaultProps = {
    autoComplete: 'on',
    expandOnFocus: false,
    allowFreeform: false,
    label: undefined,
    errorMessage: undefined,
    help: undefined,
    disabled: false
  };

  constructor(props) {
    super(props);
    this.state = {
      isCalloutVisible: false
    };
    this._onRenderLabel = this._onRenderLabel.bind(this);
    this._onClick = this._onClick.bind(this);
    this._onDismiss = this._onDismiss.bind(this);
  }

  _onRenderLabel(props) {
    const { label, help, componentId } = props;

    const styles = getLabelClassNames(props);
    let { isCalloutVisible } = this.state;
    const helpElement = isValidElement(help) ? help : <p>{help}</p>;
    return (
      <div className={styles.labelArea}>
        <span className={styles.label}>
          {label ? (
            <Label className={styles.labelText} htmlFor={componentId}>
              {label}
            </Label>
          ) : null}
        </span>
        {help && (
          <span
            className={styles.labelIconArea}
            ref={menuButton => (this._iconButtonElement = menuButton)}
          >
            <IconButton
              iconProps={{ iconName: 'HelpOutline' }}
              title="HelpOutline"
              ariaLabel="Hjelp"
              onClick={this._onClick}
              className={styles.icon}
              aria-haspopup={true}
            />
          </span>
        )}
        {isCalloutVisible && (
          <Callout
            directionalHint={Callout.POS_TOP_LEFT}
            color={Callout.HELP}
            ariaLabel={'Hjelpetekst'}
            target={this._iconButtonElement}
            onClose={this._onDismiss}
          >
            {helpElement}
          </Callout>
        )}
      </div>
    );
  }

  _onClick() {
    this.setState({
      isCalloutVisible: !this.state.isCalloutVisible
    });
  }

  _onDismiss() {
    this.setState({
      isCalloutVisible: false
    });
  }

  render() {
    const {
      children,
      errorMessage,
      label,
      autoComplete,
      allowFreeform,
      help,
      onRenderLabel,
      className,
      id,
      inputSize,
      ...props
    } = this.props;
    const labelProps = { label, help };
    return (
      <div id={id}>
        {onRenderLabel
          ? onRenderLabel(labelProps)
          : this._onRenderLabel(labelProps)}
        <SkeCombobox
          {...props}
          label={undefined}
          ariaLabel={label}
          role="combobox"
          shouldOpenOnFocus={true}
          className={classnames(getClassNames(this.props), className)}
          autoComplete={autoComplete}
          allowFreeform={allowFreeform}
          onFocus={this._basicComboBoxOnClick}
          componentRef={this._basicComboBoxComponentRef}
          calloutProps={{
            className: getOptionsClassNames(this.props)
          }}
          inputSize={inputSize}
        >
          {children}
        </SkeCombobox>
        {errorMessage && (
          <label className={getErrorClassNames(this.props)}>
            {errorMessage}
          </label>
        )}
      </div>
    );
  }

  _basicComboBoxComponentRef = component => {
    this._basicCombobox = component;
  };

  _basicComboBoxOnClick = () => {
    this._basicCombobox.focus();
    const { onFocus } = this.props;
    onFocus && onFocus();
  };
}
