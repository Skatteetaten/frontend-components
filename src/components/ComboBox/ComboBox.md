** En ComboBox ser ut og fungerer i hovedsak som en som nedtrekksliste - men har den forskjellen brukeren kan skrive i feltet. Det er spesielt nyttig hvis listen over valg er lang. **

```js
import ComboBox from '@skatteetaten/frontend-components/ComboBox';
import Grid from '@skatteetaten/frontend-components/Grid';

const initialState = {
  options: [
    { key: 'A', text: 'alfa', value: 'Alfa' },
    { key: 'B', text: 'beta', value: 'Alfa' },
    { key: 'C', text: 'gamma', value: 'Alfa' },
    { key: 'D', text: 'delta', value: 'Alfa' },
    { key: 'E', text: 'echo', value: 'Alfa' }
  ]
};

<div>
  <Grid>
    <Grid.Row>
      <Grid.Col lg={4}>
        <ComboBox
          label="Nedtrekksliste"
          help="Hjelpetekst"
          placeHolder="Velg"
          options={state.options}
          allowFreeform={false}
          ariaLabel="Eksempel ComboBox"
          useComboBoxAsMenuWidth
        />
      </Grid.Col>
      <Grid.Col lg={4}>
        <ComboBox
          label="Med autoutfylling"
          help="Feltet foreslår en verdi når du begynner å skrive"
          options={state.options}
          allowFreeform={true}
          autoComplete={'on'}
          ariaLabel="Eksempel ComboBox"
          useComboBoxAsMenuWidth
        />
      </Grid.Col>
    </Grid.Row>
  </Grid>
</div>;
```

Stor versjon:

```js
import ComboBox from '@skatteetaten/frontend-components/ComboBox';
import Grid from '@skatteetaten/frontend-components/Grid';

const initialState = {
  options: [
    { key: 'A', text: 'alfa', value: 'Alfa' },
    { key: 'B', text: 'beta', value: 'Alfa' },
    { key: 'C', text: 'gamma', value: 'Alfa' },
    { key: 'D', text: 'delta', value: 'Alfa' },
    { key: 'E', text: 'echo', value: 'Alfa' }
  ]
};

<div>
  <Grid>
    <Grid.Row>
      <Grid.Col lg={4}>
        <ComboBox
          label="Nedtrekksliste"
          help="Hjelpetekst"
          inputSize="large"
          placeHolder="Velg"
          options={state.options}
          allowFreeform={false}
          ariaLabel="Eksempel ComboBox"
          useComboBoxAsMenuWidth
        />
      </Grid.Col>
    </Grid.Row>
  </Grid>
</div>;
```

```js noeditor
import Accordion from '@skatteetaten/frontend-components/Accordion';
import AccordionItem from '@skatteetaten/frontend-components/Accordion/AccordionItem';

<Accordion>
  <AccordionItem
    isOpen
    toggleContent
    toggleButtonText={'Bruk'}
    stepId={'step-1-1'}
  >
    <p>
      ComboBoxen er en fleksibel nedtrekksliste som kan tilpasses til ulike
      sitasjoner:
    </p>
    <ul>
      <li>Kun velge eller skrive fra et sett gyldige verdier.</li>
      <li>
        Velge eller skrive blant et sett med verdier, eller skrive inn nye
      </li>
      <li>Som en nedtrekksliste, uten mulighet til å skrive</li>
      <li>Automatisk vise meny når feltet får fokus.</li>
    </ul>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Universell utforming'}
    stepId={'step-1-2'}
  >
    <ul>
      <li>Sjekk at alle funksjoner kan nås ved hjelp av tastaturet. </li>
      <li>
        Vær spesielt oppmerksom dersom man bruker <b>autoExpand</b>. Brukere av
        skjermleser forventer å aktivt aktivere elementene på skjermen, og
        dersom de åpnes automatisk kan det være vanskelig å orientere seg.
      </li>
    </ul>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Tekniske detaljer'}
    stepId={'step-1-3'}
  >
    <p>
      <a
        href="https://developer.microsoft.com/en-us/fabric#/components/ComboBox#Implementation"
        target="_blank"
      >
        Se flere tilgjengelige props i Fabric dokumentasjonen
      </a>
    </p>
  </AccordionItem>
</Accordion>;
```
