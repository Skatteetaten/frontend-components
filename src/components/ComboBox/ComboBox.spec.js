import React from 'react';
import { matches } from './../utils/test-utils';
import { shallow, mount } from 'enzyme';
import ComboBox from './ComboBox';

function oppsettShallow(props) {
  return shallow(<ComboBox {...props} />);
}

function oppsettFullDOM(props) {
  return mount(<ComboBox {...props} />);
}

describe('Combobox komponent', () => {
  it('matcher med snapshot', () => {
    matches(<ComboBox />);
  });

  it('rendrer Combobox med default props', () => {
    const wrapper = oppsettShallow();
    const combobox = wrapper.find('SkeCombobox');
    expect(combobox.prop('autoComplete')).toEqual('on');
    expect(combobox.prop('expandOnFocus')).toEqual(false);
    expect(combobox.prop('allowFreeform')).toEqual(false);
    expect(combobox.prop('label')).toEqual(undefined);
    expect(combobox.prop('disabled')).toEqual(false);
    expect(combobox.prop('shouldOpenOnFocus')).toEqual(true);
  });

  it('rendrer Cobobox med riktige props ', () => {
    const wrapper = oppsettShallow({
      id: 'Combobox-1',
      disabled: true
    });
    expect(wrapper.find('SkeCombobox').prop('disabled')).toEqual(true);
    expect(wrapper.first('div').prop('id')).toEqual('Combobox-1');
  });

  it('rendrer Cobobox label', () => {
    const wrapper = oppsettShallow({
      label: 'Nedtrekksliste'
    });
    const label = wrapper.find('StyledLabelBase');

    expect(label.html()).toContain('labelText');
    expect(label.html()).toContain('Nedtrekksliste');
  });

  it('rendrer Cobobox med hjelpetekst', () => {
    const wrapper = oppsettShallow({
      help: 'Hjelpetekst'
    });
    const helpIcon = wrapper.find('CustomizedIconButton');
    expect(helpIcon.prop('ariaLabel')).toEqual('Hjelp');
    expect(helpIcon.prop('title')).toEqual('HelpOutline');

    helpIcon.simulate('click');
    const callout = wrapper.find('Callout');
    expect(wrapper.exists('Callout')).toEqual(true);
    expect(callout.html()).toContain('Hjelpetekst');
    expect(callout.prop('color')).toEqual('lightGreen');
    expect(callout.prop('ariaLabel')).toEqual('Hjelpetekst');
  });

  it('rendrer Cobobox med hjelpetekst med markup', () => {
    const wrapper = oppsettShallow({
      help: <em>'Hjelpetekst'</em>
    });
    const helpIcon = wrapper.find('CustomizedIconButton');

    helpIcon.simulate('click');
    const callout = wrapper.find('Callout');
    expect(wrapper.exists('Callout')).toEqual(true);
    expect(callout.find('em').html()).toContain('Hjelpetekst');
  });

  it('rendrer nedtrekksliste nÃ¥r Cobobox klikkes ', () => {
    const wrapper = oppsettFullDOM({
      options: [{ key: 'A', text: 'alfa', value: 'Alfa' }]
    });

    const button = wrapper.find('CustomizedIconButton');
    button.simulate('click');
    expect(wrapper.exists('.ms-ComboBox-optionText')).toEqual(true);
  });

  it('rendrer feilmelding under Cobobox ', () => {
    const wrapper = oppsettShallow({
      errorMessage: 'Feilmelding'
    });

    expect(wrapper.html()).toContain('SkeComboBoxError');
    expect(wrapper.html()).toContain('Feilmelding');
  });
});
