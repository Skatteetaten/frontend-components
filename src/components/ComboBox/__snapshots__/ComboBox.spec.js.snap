// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Combobox komponent matcher med snapshot 1`] = `
<div>
  <div
    className="labelArea-41"
  >
    <span
      className="label-42"
    />
  </div>
  <div
    className="ms-ComboBox-container SkeCombobox-46"
    onFocus={[Function]}
    role="combobox"
  >
    <div
      className="ms-ComboBox css-48"
      id="ComboBox0wrapper"
    >
      <input
        aria-autocomplete="both"
        aria-disabled={false}
        aria-expanded={false}
        autoCapitalize="off"
        autoComplete="off"
        className="ms-ComboBox-Input css-49"
        data-is-interactable={true}
        data-lpignore={true}
        id="ComboBox0-input"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onCompositionEnd={[Function]}
        onCompositionStart={[Function]}
        onCompositionUpdate={[Function]}
        onFocus={[Function]}
        onInput={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onTouchStart={[Function]}
        readOnly={true}
        role="combobox"
        spellCheck={false}
        type="text"
        value=""
      />
      <button
        aria-hidden={true}
        className="ms-Button ms-Button--icon ms-ComboBox-CaretDown-button root-55"
        data-is-focusable={false}
        onBlur={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        role="presentation"
        tabIndex={-1}
        type="button"
      >
        <span
          className="ms-Button-flexContainer flexContainer-56"
          data-automationid="splitbuttonprimary"
        >
          <i
            aria-hidden={true}
            className="ms-Icon root-37 css-63 ms-Button-icon icon-58"
            data-icon-name="ChevronDown"
            role="presentation"
          >
            ï…€
          </i>
        </span>
      </button>
    </div>
    <div
      aria-atomic="true"
      aria-live="polite"
      className="css-50"
      id="ComboBox0-error"
      role="region"
    >
      
    </div>
  </div>
</div>
`;
